packages/dht/src/connection/WebSocket/WebSocketConnector.ts

```diff
@@ -0,0 +1,198 @@
+import { ITransport } from '../../transport/ITransport'
+import { RoutingRpcCommunicator } from '../../transport/RoutingRpcCommunicator'
+import { RemoteWebSocketConnector } from './RemoteWebSocketConnector'
+import {
+    HandshakeMessage,
+    Message,
+    MessageType,
+    PeerDescriptor,
+    WebSocketConnectionRequest,
+    WebSocketConnectionResponse
+} from '../../proto/DhtRpc'
+import { WebSocketConnectorClient } from '../../proto/DhtRpc.client'
+import { DeferredConnection } from '../DeferredConnection'
+import { TODO } from '../../types'
+import { Logger } from '../../helpers/Logger'
+import { IWebSocketConnector } from '../../proto/DhtRpc.server'
+import { ServerCallContext } from '@protobuf-ts/runtime-rpc'
+import { v4 } from 'uuid'
+
+const logger = new Logger(module)
+
+export class WebSocketConnector extends EventEmitter implements IConnectionSource, IWebSocketConnector {
+    private WESOCKET_CONNECTOR_APP_ID = "websocketconnector"
+    private rpcCommunicator: RoutingRpcCommunicator
+    private ownPeerDescriptor: PeerDescriptor | null = null
+    private canConnectFunction: (peerDescriptor: PeerDescriptor, _ip: string, port: number) => boolean
+
+    constructor(
+        private rpcTransport: ITransport,
+        fnCanConnect: (peerDescriptor: PeerDescriptor, _ip: string, port: number) => boolean
+    ) {
+        super()
+        this.canConnectFunction = fnCanConnect.bind(this)
+
+        this.rpcCommunicator = new RoutingRpcCommunicator(this.WESOCKET_CONNECTOR_APP_ID, this.rpcTransport, {
+            rpcRequestTimeout: 10000
+        })
+
+        this.requestConnection = this.requestConnection.bind(this)
+
+        this.rpcCommunicator.registerRpcMethod(
+            WebSocketConnectionRequest,
+            WebSocketConnectionResponse,
+            'requestConnection',
+            this.requestConnection
+        )
+    }
+
+    connect({ host, port, url, ownPeerDescriptor, targetPeerDescriptor }: {
+        host?: string,
+        port?: number,
+        url?: string,
+        ownPeerDescriptor?: PeerDescriptor,
+        targetPeerDescriptor?: PeerDescriptor
+    } = {}
+    ): IConnection {
+
+        if (!host && !port && !url && ownPeerDescriptor && targetPeerDescriptor) {
+            return this.requestConnectionFromPeer(ownPeerDescriptor, targetPeerDescriptor)
+        }
+        const socket = new ClientWebSocket()
+
+        socket.once(ConnectionEvent.CONNECTED, () => {
+            this.emit(ConnectionSourceEvent.CONNECTED, socket)
+        })
+
+        let address = ''
+        if (url) {
+            address = url
+        }
+        else if (host && port) {
+            address = 'ws://' + host + ':' + port
+        }
+
+        socket.connect(address)
+        return socket
+    }
+
+    connectAsync({ host, port, url, timeoutMs }:
+        { host?: string; port?: number; url?: string; timeoutMs: number } = { timeoutMs: 1000 }): Promise<IConnection> {
+
+        return new Promise((resolve, reject) => {
```

Comment:
The enhanced `waitForEvent`/`withTimeout` is now in main (https://github.com/streamr-dev/network-monorepo/pull/829).

If we want to implement this with the helper methods it would look like this:

```
const connectAsync = async (timeout: number): Promise<ClientWebSocket> => {
    const socket = new ClientWebSocket()

    const abortController = new AbortController()
    const waitForConnection = withTimeout(
        Promise.race([
            waitForEvent(socket as any, 'connected', timeout, undefined, abortController),
            (async () => {
                await waitForEvent(socket as any, 'error', timeout, undefined, abortController)
                throw new ConnectionFailed('Could not open WebSocket connection')
            })()
        ]),
        timeout,
        new ConnectionFailed('WebSocket connection timed out'),
        abortController
    )

    socket.connect()
    await waitForConnection
    return socket
}
```

That would take to account that the convention in DHT is to throw DHT-specific errors instead of low level errors (e.g. `ConnectionFailed` instead of `TimeoutError`). That's why it would maybe make sense to have a separate helper method to wrap the convention logic:

```
import { TimeoutError, waitForEvent, withTimeout as _withTimeout } from '@streamr/utils'
export const withTimeout = async <T>(
    task: Promise<T>,
    timeoutInMs: number,
    timeoutError: Error,
    abortController?: AbortController
): Promise<T> => {
    try {
        return await _withTimeout(task, timeoutInMs, undefined, abortController)
    } catch (e) {
        if (e instanceof TimeoutError) {
            throw timeoutError
        } else {
            throw e
        }
    }
}
```

But yeah, my original suggestion isn't maybe as applicable as it was on June. Now this methods uses DHT-specific error and therefore we need to convert low level errors to DHT-specific errors. That makes the helper methods less usable.

Maybe this `waitForEvent`/`withTimeout` style would be bit more readable/maintainable compared to manual bookkeeping. But the difference is not very significant anymore.